stratum/hal/lib/tdi/es2k/es2k_port_manager.cc

```diff
@@ -0,0 +1,245 @@
+
+Es2kPortManager::Es2kPortManager() : port_status_event_writer_(nullptr) {}
+
+namespace {
+
+::util::StatusOr<int> AutonegHalToBf(TriState autoneg) {
+  switch (autoneg) {
+    case TRI_STATE_UNKNOWN:
+      return 0;
+    case TRI_STATE_TRUE:
+      return 1;
+    case TRI_STATE_FALSE:
+      return 2;
+    default:
+      RETURN_ERROR(ERR_INVALID_PARAM) << "Invalid autoneg state.";
+  }
+}
+
+// A callback function executed in SDE port state change thread context.
+bf_status_t sde_port_status_callback(
+    bf_dev_id_t device, bf_dev_port_t dev_port, bool up, void* cookie) {
+  absl::Time timestamp = absl::Now();
+  Es2kPortManager* es2k_port_manager = Es2kPortManager::GetSingleton();
+  if (!es2k_port_manager) {
+    LOG(ERROR) << "Es2kPortManager singleton instance is not initialized.";
+    return BF_INTERNAL_ERROR;
+  }
+  // Forward the event.
+  auto status =
+      es2k_port_manager->OnPortStatusEvent(device, dev_port, up, timestamp);
+
+  return status.ok() ? BF_SUCCESS : BF_INTERNAL_ERROR;
+}
+
+}  // namespace
+
+Es2kPortManager* Es2kPortManager::CreateSingleton() {
+  absl::WriterMutexLock l(&init_lock_);
+  if (!singleton_) {
+    singleton_ = new Es2kPortManager();
+  }
+
+  return singleton_;
+}
+
+Es2kPortManager* Es2kPortManager::GetSingleton() {
+  absl::ReaderMutexLock l(&init_lock_);
+  return singleton_;
+}
+
+::util::StatusOr<PortState> Es2kPortManager::GetPortState(int device, int port) {
+  // Unsupported. Returns PORT_STATE_DOWN if called.
+  return PORT_STATE_DOWN;
+}
+
+::util::Status Es2kPortManager::GetPortCounters(
+    int device, int port, PortCounters* counters) {
+  uint64_t stats[BF_PORT_NUM_COUNTERS] = {0};
+
+  return ::util::OkStatus();
+}
+
+::util::Status Es2kPortManager::OnPortStatusEvent(
+    int device, int port, bool up, absl::Time timestamp) {
+  // Create PortStatusEvent message.
+  PortState state = up ? PORT_STATE_UP : PORT_STATE_DOWN;
+  PortStatusEvent event = {device, port, state, timestamp};
+
+  {
+    absl::ReaderMutexLock l(&port_status_event_writer_lock_);
+    if (!port_status_event_writer_) {
+      return ::util::OkStatus();
+    }
+    return port_status_event_writer_->Write(event, kWriteTimeout);
+  }
+}
+
+::util::Status Es2kPortManager::RegisterPortStatusEventWriter(
+    std::unique_ptr<ChannelWriter<PortStatusEvent>> writer) {
+  absl::WriterMutexLock l(&port_status_event_writer_lock_);
+  port_status_event_writer_ = std::move(writer);
+  return ::util::OkStatus();
+}
+
+::util::Status Es2kPortManager::UnregisterPortStatusEventWriter() {
+  absl::WriterMutexLock l(&port_status_event_writer_lock_);
+  port_status_event_writer_ = nullptr;
+  return ::util::OkStatus();
+}
+
+::util::Status Es2kPortManager::GetPortInfo(
+    int device, int port, TargetDatapathId *target_dp_id) {
+  return ::util::OkStatus();
+}
+
+// TODO: Check with Sandeep which Add port is applicable or do we really
+// need it since we don't add port for MEV
```

Comment:
We don't add port for ES2K. we can keep the function and return OK. Function on line 146 can be removed for ES2K